### PR TITLE
fix: correct the args for `disambiguate the associated function` diagnostic

### LINKED ISSUE #118819: `disambiguate the associated function` diagnostic take receiver as first arg mistakely
### Code

```Rust
struct A {
}

trait M {
    fn new();
}
impl M for A {
    fn new() {
        todo!()
    }
}

trait N {
    fn new();
}
impl N for A {
    fn new() {
        todo!()
    }
}

fn main() {
    let a = A {};
    a.new();
}
```

### Current output

```Shell
error[E0599]: no method named `new` found for struct `A` in the current scope
  --> src/main.rs:37:7
   |
14 | struct A {
   | -------- method `new` not found for this struct
...
37 |     a.new();
   |       ^^^ this is an associated function, not a method
   |
   = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: candidate #1 is defined in the trait `M`
  --> src/main.rs:18:5
   |
18 |     fn new();
   |     ^^^^^^^^^
note: candidate #2 is defined in the trait `N`
  --> src/main.rs:27:5
   |
27 |     fn new();
   |     ^^^^^^^^^
help: disambiguate the associated function for candidate #1
   |
37 |     <A as M>::new(a);
   |     ~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
   |
37 |     <A as N>::new(a);
   |     ~~~~~~~~~~~~~~~~
```

### Desired output

```Shell
help: disambiguate the associated function for candidate #1
   |
37 |     <A as M>::new();
   |     ~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
   |
37 |     <A as N>::new();
   |     ~~~~~~~~~~~~~~~~
```

### Rationale and extra context

This is somehow silimar to https://github.com/rust-lang/rust/issues/118469, we shouldn't take receiver as first arg all the cases.

### Other cases

_No response_

### Anything else?

rustc --version
rustc 1.76.0-nightly (d86d65bbc 2023-12-10)

### PR DESCRIPTION
**Pull request by Young-Flash**
This is somehow silimar to https://github.com/rust-lang/rust/pull/118502, we shouldn't take receiver as first arg all the cases.

close https://github.com/rust-lang/rust/issues/118819

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -3324,19 +3332,22 @@ fn print_disambiguation_help<'tcx>(
...
-                        .sess
-                        .source_map()
-                        .span_to_snippet(arg.span)
-                        .unwrap_or_else(|_| { "_".to_owned() }))
-                    .collect::<Vec<_>>()
-                    .join(", "),
-            );
+            // If the type of first arg of this assoc function is `Self` or current trait impl type, we need to take the receiver as args. Otherwise, we don't.
+            let args = if let Some(t) = first_arg_type
+                && (t.to_string() == "Self" || t == trait_impl_type)
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3337]: Without having checked if the rest of your implementation makes sense, I'll suggest the following:

```suggestion
                && (t == tcx.types.self_param || t == trait_impl_type)
```

**Comment by compiler-errors** [comment]
r? fmease

**Comment by Young-Flash** [comment]
can't figure out why CI failed with the following, but it compile on my local. any hints? @fmease @compiler-errors 
```txt
error[E0599]: no method named `find_by_def_id` found for struct `rustc_middle::hir::map::Map` in the current scope
    --> compiler/rustc_hir_typeck/src/method/suggest.rs:3356:61
     |
3356 |                         Some(local_def_id) => match hir_map.find_by_def_id(local_def_id) {
     |                                                             ^^^^^^^^^^^^^^ method not found in `Map<'_>`
```

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 2]
```diff
@@ -3324,6 +3325,13 @@ fn print_disambiguation_help<'tcx>(
         {
             let def_kind_descr = tcx.def_kind_descr(item.kind.as_def_kind(), item.def_id);
             let item_name = item.ident(tcx);
+            let first_arg_type = tcx
+                .fn_sig(item.def_id)
+                .instantiate_identity()
+                .skip_binder()
+                .inputs()
+                .get(0)
+                .and_then(|first| Some(first.peel_refs()));
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3334]: This makes sense but it's prone to errors:

```
trait A { fn f(self); }
trait B { fn f(self); }

#[derive(Clone, Copy)]
struct S;

impl A for S { fn f(self) {} }
impl B for S { fn f(self) {} }

fn main() {
    (&&&S).f(); // Here, we suggest `A::f((&&&S))` but we should suggest `A::f(***S)`
}

```

Doesn't need to be fixed in this PR just making note of it.

Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3334]:
```suggestion
                .map(|first| first.peel_refs());
```

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 3]
```diff
@@ -3332,19 +3340,56 @@ fn print_disambiguation_help<'tcx>(
                 .get(0)
                 .and_then(|ty| ty.ref_mutability())
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3341]: The amount of references is not kept in sync between `first_arg_type` and `rcvr_ref` here:

```rs
trait A { fn f(self: &mut &Self); }
trait B { fn f(self: &mut &Self); }

struct S;

impl A for S { fn f(self: &mut &Self) {} }
impl B for S { fn f(self: &mut &Self) {} }

fn main() {
    (&mut &S).f(); // We suggest `A::f(&mut (&mut &S))` but it should be `A::f(&mut &S);`
}

```

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 4]
```diff
@@ -3324,6 +3325,13 @@ fn print_disambiguation_help<'tcx>(
         {
             let def_kind_descr = tcx.def_kind_descr(item.kind.as_def_kind(), item.def_id);
             let item_name = item.ident(tcx);
+            let first_arg_type = tcx
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:3328]: Could you merge the two statements? Something like `let (first_arg_type, recv_ref) = …`? They share a lot.

**Review by fmease — CHANGES_REQUESTED** [review]
Sorry for taking so long! I have some style nits and some suggestions. 

This diagnostic generally has quite a few false positives and negatives before and after your changes, so I guess it's not worth making it perfect in this PR. I have some suggestions like trying out `can_eq`. If you can't make it work, then that's fine, too.

**Comment by Young-Flash** [comment]
sorry for the slow response and thank you for taking the time to review this. I make some modification and leave the rest for future work (so this PR could be focused), LMK if there are somethings can improve

**Review by fmease — APPROVED** [review]
Thanks, for implementing my feedback, it looks good! Further improvements can potentially be made in follow-up PRs.

**Pull request merged**